### PR TITLE
Tooling makeover

### PR DESCRIPTION
**Pull request by safl**
Motivation and progress of this PR is tracked here:

https://github.com/OpenMPDK/xNVMe/issues/146

Overview of Changes
================

This PR is gigantic with 6000 lines removed and 3000 added. However, only about 500 lines are actual interesting logic, the rest is boiler-plate code of glue-tests "ported" from **cijoe/Bash** to **cijoe/Python/pytest**.

The interesting parts are:

```
toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py
toolbox/workflow/*
.github/workflows/verify.yml
```
See the commit-messages for additional descriptions and have a look at the reports provided as artifacts:
https://github.com/OpenMPDK/xNVMe/pull/148/checks

It is functionally equivalent to the previous infra with the following thing **todo**:
* Port hook for fabrics-setup (should be part of the driver-magic in ``conftest.py``)

Any input on glue-tests missing etc. is most welcome.

Addition
======

Integrated changes on this PR from: https://github.com/OpenMPDK/xNVMe/pull/149
Since this PR had to make other changes to disable the FreeBSD build, so instead the fix to build is added here fixing  https://github.com/OpenMPDK/xNVMe/issues/150

### CODE REVIEW

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_scc.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+"""
+This is a port of the tests:
+xnvme_tests_scc_idfy.sh                 --> test_idfy()
+xnvme_tests_scc_scopy_async.sh          --> test_scopy()
+xnvme_tests_scc_scopy_sync.sh           --> test_scopy_clear()
+xnvme_tests_scc_scopy_msrc_async.sh     --> test_scopy_msrc()
+xnvme_tests_scc_scopy_msrc_sync.sh      --> test_scopy_msrc_clear()
+xnvme_tests_scc_support.sh              --> test_support()
+
+Note: the tests were named "async" did not seem to have any async be-instrumentation.
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_scc.py:10]: This line is a bit confusing, I think this is an improvement
`Note: the tests named "async" did not seem to have any async be-instrumentation`

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_xnvme_file.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+"""
+New addition
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_xnvme_file.py:2]: This does not seem to be very relevant.
I think it would be better to add a description of the test

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+xnvme_idfy_ctrlr.sh   --> test_idfy_ctrlr()
+xnvme_idfy_ns.sh      --> test_idfy_ns()
+xnvme_info.sh         --> test_info()
+xnvme_library_info.sh --> test_library_info()
+xnvme_log-erri.sh     --> test_log_erri()
+xnvme_log-health.sh   --> test_log_health()
+xnvme_log.sh          --> test_log()
+xnvme_padc.sh         --> test_padc()
+xnvme_pioc.sh         --> test_pioc()
+xnvme_sanitize.sh     --> MISSING
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:20]: This line should be removed

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+
+Observation:
+    How was idfy-cs tested previously?
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:23]: I don't think this should be here

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    err, _ = cijoe.run(f"xnvme format {cli_args}")
+    assert not err
+
+
+@pytest.mark.parametrize(
+    "device,be_opts,cli_args",
+    xnvme_setup(labels=["nvm"], opts=["be", "admin"]),
+    indirect=["device"],
+)
+def test_format(cijoe, device, be_opts, cli_args):
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:135]: This test should be renamed to test_sanitize

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+
+    pytest.skip(reason="TODO: always fails. Investigate.")
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:137]: This should be investigated before merging

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_zoned.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+zoned_idfy_ns.sh                --> test_idfy_ns()
+zoned_info.sh                   --> test_info()
+zoned_mgmt_open.sh              --> MISSING
+zoned_read.sh                   --> test_read()
+zoned_report.sh                 --> test_report_limit()
+zoned_report_all.sh             --> test_report_all()
+zoned_report_one.sh             --> test_report_single()
+zoned_report_some.sh            --> test_report_some()
+zoned_write.sh                  --> test_reset_report_write_report()
+zoned_mgmt_open.sh              --> test_reset_open_report()
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_zoned.py:17]: This line should be moved up to line 10 for the alphabetical order to make sense

**Review by karlowich — CHANGES_REQUESTED** [review]
Looks good!
I have provided nit-picks

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_ioworker.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+from cijoe.xnvme.tests.conftest import xnvme_device_driver as device
+from cijoe.xnvme.tests.conftest import xnvme_setup
+
+
+@pytest.mark.parametrize(
+    "device,be_opts,cli_args",
+    xnvme_setup(labels=["dev"], opts=["be", "sync", "async", "admin"]),
+    indirect=["device"],
+)
+def test_verify(cijoe, device, be_opts, cli_args):
```
Comment (karlowich) [inline toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_ioworker.py:14]: This needs to be expanded with a `test_verify_sync` that replaces the `xnvme_tests_ioworker_sync.sh` from `next`

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
+"""
+This is a port of the tests:
+
+xnvme_enum.sh         --> test_enum()
+xnvme_enum_fabrics.sh --> test_enum_fabrics()
+xnvme_feature_get.sh  --> test_feature_get()
+xnvme_feature_set.sh  --> test_feature_set()
+xnvme_format.sh       --> test_format()
+xnvme_sanitize.sh     --> test_sanitize() -- TODO: Needs some investigation
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:9]: I went with removing all of these comments and adding them to the commit-message instead.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+
+@pytest.mark.parametrize(
+    "device,be_opts,cli_args",
+    xnvme_setup(labels=["dev"], opts=["be", "admin", "sync", "async"]),
+    indirect=["device"],
+)
+def test_fio_engine(cijoe, device, be_opts, cli_args, request):
+    """
+    The construction of the fio-invokation is done in 'cijoe.fio.wrapper.fio_fancy'
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:21]: fixed.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_zoned.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
+"""
+This is a port of the tests:
+
+zoned_append.sh                 --> test_append()
+zoned_changes.sh                --> test_changes()
+zoned_enum.sh                   --> test_enum()
+zoned_idfy_ctrlr.sh             --> test_idfy_ctrlr()
+zoned_idfy_ns.sh                --> test_idfy_ns()
+zoned_info.sh                   --> test_info()
+zoned_mgmt_open.sh              --> MISSING
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_zoned.py:10]: I went with removing all of these comments and adding them to the commit-message instead.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py (hunk 3 above)
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:20]: I went with removing all of these comments and adding them to the commit-message instead.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py (hunk 4 above)
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:23]: I went with removing all of these comments and adding them to the commit-message instead.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py (hunk 6 above)
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme.py:137]: Made a note of this, post-poning fixage.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_zoned.py (hunk 7 above)
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_zoned.py:17]: I went with removing all of these comments and adding them to the commit-message instead.

**Comment by safl** [comment]
@karlowich @mads Thanks for the feedback here and offline.
I have revised, when the CI passed, I'll go ahead and squash, then run again.
Let me know if there is anything, thanks!

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_async_intf.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+xnvme_tests_async_intf02.sh     --> test_init_term()
+xnvme_tests_async_intf03.sh     --> test_init_term()
+xnvme_tests_async_intf04.sh     --> test_init_term()
+
+The difference of the above tests are the variance of the parameter --count.
+The first three had hard-coded 1... 4 ... 8.
+The last looped over the sequence [1, ..., 128]
+
+The first three are redundant as the general form catches all.
+Thus, just implementing that. Also, this happended due to issue previously. Now, it just
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_async_intf.py]: removed / fixed.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_cli.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
+"""
+New addition
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_cli.py]: removed / fixed.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_enum.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+from cijoe.xnvme.tests.conftest import xnvme_setup
+
+
+@pytest.mark.parametrize(
+    "device,be_opts,cli_args",
+    xnvme_setup(labels=["dev"], opts=["be"]),
+    indirect=["device"],
+)
+def test_open(cijoe, device, be_opts, cli_args):
+    pytest.skip(reason="Currently not implemented, comments this out after merge.")
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_enum.py:9]: Added to future fixing.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_ioworker.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
+"""
+New addition
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_ioworker.py]: removed / fixed.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_ioworker.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    err, _ = cijoe.run(f"xnvme_tests_ioworker verify {cli_args}")
+    assert not err
+
+
+@pytest.mark.parametrize(
+    "device,be_opts,cli_args",
+    xnvme_setup(labels=["bdev"], opts=["be", "sync", "async", "admin"]),
+    indirect=["device"],
+)
+@pytest.mark.skip(reason="FIXME: --direct=1 needs investigation")
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_ioworker.py:17]: It is not a regression, since we haven't done this before.
So I would rather go ahead and get the infra in and then use it to speed of the fixing of this and other issues.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_scc.py (hunk 1 above)
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_scc.py]: removed / fixed.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_xnvme_file.py (hunk 2 above)
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_xnvme_file.py]: removed / fixed.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py [hunk 17]
```diff
@@ -0,0 +1,10 @@
+import pytest
+from cijoe.xnvme.tests.conftest import xnvme_device_driver as device
+from cijoe.xnvme.tests.conftest import xnvme_parametrize
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py:3]: I recall having issues with that, which i think is related to:
https://github.com/pytest-dev/pytest/issues/1596

However, since the intent is enabling easy deployment of the tests and running them, via "pytest --pyargs cijoe.xnvme.tests". Not so much the inplace-edit-and-run. Then i didnt pursue it any further.

I don't see saving a single line of import atatement per test-module as a reason to not go ahead with merging. Wouldn't you agree?

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py:3]: To elaborate a bit on this, for in-place-edit-and run, one can set the ``PYTHONPATH``, for example:

	PYTHONPATH="$HOME/git/xnvme/toolbox/cijoe-pkg-xnvme/src" pytest --pyargs cijoe.xnvme.tests -k info ...

Then Python will find the non-installed package before any installed packages. Doing so enables the in-place-edit-and-run. The above can also be used, in case one just does not want to install the package.

Another thing that I have found myself doing is, writing a test, say ``test_foo.py``:

	from cijoe.xnvme.tests.conftest import xnvme_device_driver as device
	from cijoe.xnvme.tests.conftest import xnvme_parametrize

	@xnvme_parametrize(labels=[], opts=["be", "sync", "async"])
	def test_foo(cijoe, device, be_opts, cli_args):
	    assert True

And having that test in an arbitrary "out-of-tree" location, that is, not a child of "cijoe-pkg-xnvme/src/cijoe/xnvme/tests". And running it:

	pytest test_foo.py ...

Since it is "out-of-tree" the fully qualified imports are needed. Once I am done with the test I copy it into the python-package without any need to change the imports. Similarly, one can copy any of the tests "in-tree" to an arbitrary "out-of-tree" location and run it "standalone" without changing the imports.

So, to address whether the fully-qualified import paths should be made relative, I would prefer not to, because:

* They can be run 'in-place' by setting the PYTHONPATH
* Tests can be moved in and out of "tree" without having to change the imports

I find the latter really convenient.

Regarding the import of "xnvme_device_driver as device", i agree that it would be nice if it could be avoided. However, it seems like a deficiency of pytest-collection, not really something we can affect without making other sacrifices.

[the remark this answers is not in the document]
Comment (Baekalfen) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py:3]: > the intent is enabling easy deployment of the tests and running them, via "pytest --pyargs cijoe.xnvme.tests". Not so much the inplace-edit-and-run.

In what cases would we install the package instead of running it from the source? When developing, and deploying to GHA, it would seem easiest not to have to build the package and install it first. And what I'm proposing won't block you from doing it anyway.

> I don't see saving a single line of import atatement per test-module as a reason to not go ahead with merging. Wouldn't you agree?

It's not about have an import statement more or less, it's the underlaying issue that concerns me. We need to find the root cause and either say we accept it, or fix it.


> one can set the PYTHONPATH

That is a possible workaround, but it also seems like a lot of added complexity, for something that should just work.

> having that test in an arbitrary "out-of-tree" location, that is, not a child of "cijoe-pkg-xnvme/src/cijoe/xnvme/tests".

Is that a plausible scenario? Wouldn't you be developing the tests and store them all in the same place? Or what would the benefit of that be?

> However, it seems like a deficiency of pytest-collection

That again seems to be a symptom of the root cause I want to avoid. The `conftest.py` is not properly located during collection. It should be an easy thing to fix.

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py:3]: > > the intent is enabling easy deployment of the tests and running them, via "pytest --pyargs cijoe.xnvme.tests". Not so much the inplace-edit-and-run.
> 
> In what cases would we install the package instead of running it from the source? When developing, and deploying to GHA, it would seem easiest not to have to build the package and install it first. And what I'm proposing won't block you from doing it anyway.
> 

This is how it is being used now. It is fairly convenient to be able to do ``pytest --pyargs cijoe.xnvme.tests``  instead of relying on the availability of the source repository and knowing where it is located. This usage takes care of it. 

> > I don't see saving a single line of import atatement per test-module as a reason to not go ahead with merging. Wouldn't you agree?
> 
> It's not about have an import statement more or less, it's the underlaying issue that concerns me. We need to find the root cause and either say we accept it, or fix it.
> 

As mentioned earlier, it seems to me that it is a long-lived pytest-deficiency https://github.com/pytest-dev/pytest/issues/1596
Going back to 2016. I am fine with it, the functionality introduced with this PR works as intended and it is a vast improvement on the status quo.

> > one can set the PYTHONPATH
> 
> That is a possible workaround, but it also seems like a lot of added complexity, for something that should just work.
> 
> > having that test in an arbitrary "out-of-tree" location, that is, not a child of "cijoe-pkg-xnvme/src/cijoe/xnvme/tests".
> 
> Is that a plausible scenario? Wouldn't you be developing the tests and store them all in the same place? Or what would the benefit of that be?
> 

I have been doing it, that is why I mention it.

> > However, it seems like a deficiency of pytest-collection
> 
> That again seems to be a symptom of the root cause I want to avoid. The `conftest.py` is not properly located during collection. It should be an easy thing to fix.

If it is an easy thing to fix, then please send a PR after this fixing it. Regardless, the only impact it will have on the functionality of this PR is whether or not those import statements are needed.

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py:3]: Here is another perspective, before this PR there was **no** way of running the tests without installing the package.
At least now, there **is** a way, by setting ``PYTHONPATH`` and running them directly from the repository.

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py]: @Baekalfen thanks for the offline discussion it helped me understand your perspective better.

If you have the time, then have a look a the ``revised: ...`` commit. Briefly:

* Imports are now relative instead of absolute.

* Inside the conf-test the ``xnvme_device_driver()`` is now renamed to
``device()`` to satisfy the pytest.parametrize/indirect. Thus the
``xnvme_device_driver as device`` import is no longer needed and removed.

* One can now run the tests in-place **without** having to specify ``PYTHONPATH`` and most importantly **without** having to install the package.

* Running it as a package still works.

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py]: Also, just added a brief description on how to run pytest directly, that is, without the use of **cijoe**.

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/examples/test_xnvme_hello.py]: Note, "revised" are gone as the history has been cleaned up.

File: subprojects/spdk.wrap [hunk 18]
```diff
@@ -1,6 +1,6 @@
 [wrap-git]
 url = https://github.com/spdk/spdk.git
-revision = v21.10
+revision = v22.05
```
Comment (karlowich) [inline subprojects/spdk.wrap:3]: I think the change from SPDK 21 to SPDK 22 should be a separate PR, I think it is worth it to make #148 as simple as possible

**Review by karlowich — CHANGES_REQUESTED** [review]
Here is my second batch of nitpicks.
They're all in the category of nice-to-have, but also easy to fix

File: subprojects/spdk.wrap (hunk 18 above)
Comment (karlowich) [reply subprojects/spdk.wrap:3]: I realize now that it was originally a separate PR (#149), and that was based on this.
Maybe it should be the other way around?

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+    The driver managing for an NVMe device is in Linux attached to one of:
+
+    * nvme (Kernel managed NVMe driver)
+    * vfio-pci (User space)
+    * uio-generic (User space)
+
+    Switching attachment is done via the 'xnvme-driver' cli-tool.
+
+    This class encapsulates the cli-tool via static helpers (kernel_detach,
+    kernal_attach). Whether or not a device needs one or the other is handlded
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:167]: ack.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 20]
```diff
@@ -0,0 +1,276 @@
+"""
+    This is the test-configuration for xNVMe
+
+    xnvme_be_opts(): helper-function used by xnvme_setup()
+
+    xnvme_setup(): generates a list of sensible backend configurations and matches
+    them up with devices from the CIJOE configuration. Emitting (device, be_opts) and
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:7]: ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 21]
```diff
@@ -0,0 +1,276 @@
...
+    @staticmethod
+    def kernel_detach(cijoe):
+        """Detach from kernel"""
+
+        cijoe.run("xnvme-driver")
+        XnvmeDriver.IS_KERNEL_ATTACHED = False
+
+    @staticmethod
+    def kernel_attach(cijoe):
+        """Detach from kernel"""
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:196]: ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 22]
```diff
@@ -0,0 +1,276 @@
...
+    """
+
+    XnvmeDriver.attach(cijoe, request.param)
+
+    return request.param
+
+
+def xnvme_parametrize(labels, opts):
+    """
+    This decorator providing all the pytest-parametrization magic in one.
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:247]: ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_cli.py [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+
+    err, _ = cijoe.run("xnvme_tests_cli optional")
+    assert not err
+
+
+@xnvme_parametrize(labels=[], opts=["be", "mem", "sync", "async", "admin"])
+def test_optional_all(cijoe, device, be_opts, cli_args):
+
+    err, _ = cijoe.run(
+        f"xnvme_tests_cli optional "
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_cli.py:14]: It is not needed here.
The unused ``cli_args`` is not pretty, however, it is provided since ``xnvme_parametrize()`` produces the list of ``(device, be_opts, cli_args)``, if the argument is removed from the test-func, then it fails.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py [hunk 24]
```diff
@@ -0,0 +1,17 @@
+from pathlib import Path
+
+from cijoe.fio.wrapper import fio_fancy
+
+from ..conftest import xnvme_parametrize
+
+
+@xnvme_parametrize(labels=["dev"], opts=["be", "admin", "sync", "async"])
+def test_fio_engine(cijoe, device, be_opts, cli_args):
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: True, the unused ``cli_args`` is not pretty, however, it is provided since ``xnvme_parametrize()`` produces the list of ``(device, be_opts, cli_args)``, if the argument is removed from the test-func, then it fails.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+
+    Usage example:
+
+      from cijoe.xnvme.tests.config import xnvme_device_driver as device
+
+      pytest.mark.parametrize("device,be_opts", {data}, indirect=["device"])
+      test_foo(cijoe, device, be_opts):
+        ...
+
+    By doing so, then each generated/parameter for the test-case is sent by this
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:231]: ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 26]
```diff
@@ -0,0 +1,276 @@
...
+    fixture. This allows the fixture to invoke device-driver attachment according the
+    the given device. This avoid having to call:
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:233]: ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_xnvme_file.py [hunk 27]
```diff
@@ -0,0 +1,21 @@
+import pytest
+
+from ..conftest import xnvme_setup_device
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/logic/test_xnvme_tests_xnvme_file.py:6]: No, since ``xnvme_parametrize()`` produces backend configurations with a usable device. In this test only the device (file) is relevant, not the backend configuration.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_lblk.py [hunk 28]
```diff
@@ -0,0 +1,49 @@
+from ..conftest import xnvme_parametrize
+
+
+def test_enum(cijoe):
+
+    err, _ = cijoe.run("lblk enum")
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_lblk.py:6]: ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme_file.py [hunk 29]
```diff
@@ -0,0 +1,21 @@
+from ..conftest import xnvme_parametrize
+
+
+@xnvme_parametrize(labels=["file"], opts=["be", "sync"])
+def test_copy_sync(cijoe, device, be_opts, cli_args):
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/tools/test_xnvme_file.py:5]: True, the unused ``cli_args`` is not pretty, however, it is provided since ``xnvme_parametrize()`` produces the list of ``(device, be_opts, cli_args)``, if the argument is removed from the test-func, then it fails.

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (karlowich) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: Okay in that case, then `cli_args` are missing from the tests in `test_cython.py`

File: subprojects/spdk.wrap (hunk 18 above)
Comment (safl) [reply subprojects/spdk.wrap:3]: The upgrade to v22.05 requires the "tooling makeover".
Moved the commit to front.

I guess I should have left the commit at PR #149 but now that it is here and were just nit-picking, how about we just leave it be?

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 30]
```diff
@@ -0,0 +1,276 @@
...
+    Usage example:
+
+      from cijoe.xnvme.tests.config import xnvme_device_driver as device
+
+      pytest.mark.parametrize("device,be_opts", {data}, indirect=["device"])
+      test_foo(cijoe, device, be_opts):
+        ...
+
+    By doing so, then each generated/parameter for the test-case is sent by this
+    fixture. This allows the fixture to invoke device-driver attachment according the
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:232]: Ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py [hunk 31]
```diff
@@ -0,0 +1,276 @@
...
+
+        driver_attachment: "kernel"
+
+    or
+
+        driver_attachment: "userspace"
+
+    When no "driver_attachment" key is provided, "kernel" is assumed.
+
+    Additionally, since swithing device-driver-attachment takes a non-trivial amount of
```
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:178]: Ack

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: Oh, i might be wrong then! Let me just check

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: Hmm, I am seeing ``function uses no argument 'cli_args'``, when unused is left out.

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: Ohh, it is skipped, that is why it is not failing with missing argument...

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: Fixed!

[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: And now, style-issue fixed.

File: subprojects/spdk.wrap (hunk 18 above)
Comment (karlowich) [reply subprojects/spdk.wrap:3]: Sure, that is fine by me

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py (hunk 31 above)
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/conftest.py:178]: DONE

File: toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (safl) [reply toolbox/cijoe-pkg-xnvme/src/cijoe/xnvme/tests/apps/test_fio.py:9]: And, now the tests ought to pass.

**Review by Baekalfen — APPROVED** [review]
Beautiful!

**Review by karlowich — APPROVED** [review]
This is great!
